tmail-backend/mailbox/encrypted/memory/src/main/java/com/linagora/tmail/encrypted/InMemoryEncryptedEmailContentStore.scala

```diff
@@ -0,0 +1,83 @@
+import reactor.core.scala.publisher.SMono
+
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.atomic.AtomicInteger
+import javax.inject.Inject
+
+case class EncryptedEmailContentStoreMemoryModule() extends AbstractModule {
+  override def configure(): Unit = {
+    bind(classOf[InMemoryEncryptedEmailContentStore]).in(Scopes.SINGLETON)
+
+    bind(classOf[EncryptedEmailContentStore]).to(classOf[InMemoryEncryptedEmailContentStore])
+  }
+}
+
+object InMemoryEncryptedEmailContentStore {
+  val emailContentStore: scala.collection.concurrent.Map[MessageId, EncryptedEmailContent] = scala.collection.concurrent.TrieMap()
+  val messageIdBlobIdStore: scala.collection.concurrent.Map[MessageId, Map[Int, BlobId]] = scala.collection.concurrent.TrieMap()
+}
+
+class InMemoryEncryptedEmailContentStore @Inject()(blobStore: BlobStore,
+                                                   bucketName: BucketName,
+                                                   storagePolicy: StoragePolicy) extends EncryptedEmailContentStore {
+
+  override def store(messageId: MessageId, encryptedEmailContent: EncryptedEmailContent): Unit = {
+    storeAttachment(messageId, encryptedEmailContent.encryptedAttachmentContents)
+    emailContentStore.put(messageId, encryptedEmailContent)
+  }
+
+  override def delete(messageId: MessageId): Unit = {
+    removeAttachment(messageId)
+    emailContentStore.remove(messageId)
+  }
+
+  override def retrieveFastView(messageId: MessageId): EncryptedEmailFastView =
+    emailContentStore.get(messageId)
+      .map(encryptedEmailContent => EncryptedEmailFastView.from(encryptedEmailContent))
+      .getOrElse(throw MessageNotFoundException(messageId))
+
+  override def retrieveDetailedView(messageId: MessageId): EncryptedEmailDetailedView =
+    emailContentStore.get(messageId)
+      .map(encryptedEmailContent => EncryptedEmailDetailedView.from(encryptedEmailContent))
+      .getOrElse(throw MessageNotFoundException(messageId))
+
+  override def retrieveAttachmentContent(messageId: MessageId, position: Int): BlobId = {
+    messageIdBlobIdStore.get(messageId)
+      .flatMap(positionBlobIdMap => positionBlobIdMap.get(position))
+      .getOrElse(throw AttachmentNotFoundException(messageId, position))
+  }
+
+  private def storeAttachment(messageId: MessageId, encryptedAttachmentContents: List[String]): Unit = {
+    Preconditions.checkNotNull(encryptedAttachmentContents)
+    if (encryptedAttachmentContents.nonEmpty) {
+      val position: AtomicInteger = new AtomicInteger(POSITION_NUMBER_START_AT)
+      val positionBlobIdMap: Map[Int, BlobId] = encryptedAttachmentContents.map(attachment => {
+        val blobId: BlobId =
+          SMono.fromPublisher(blobStore.save(bucketName, attachment.getBytes(StandardCharsets.UTF_8), storagePolicy))
+            .block()
+        position.getAndIncrement() -> blobId
+      }).toMap
+      messageIdBlobIdStore.put(messageId, positionBlobIdMap)
```

Comment:
I'm not sure to understand how you generate things here with your blobId... I thought as well, as stated in the previous ticket, that blobId for TMail should be something like `encryptedAttachment_{messageId}_{position}` ?

Reply:
I just want to save `position`- `blobId` map in memory for the contract. We need database storage for `retrieveAttachmentContent` method.